## **KNN ALGORITHM FULL IMPLEMENTATION WITHOUT SKLEARN LIBRARY**

This is the complete implementation of the K - Nearest Neighbour algorithm using the iris dataset as downloaded from the kaggle website. 

KNN algorithm is a lazy classification algorithm that classifies the test data on the basis of K nearest neighbours. The Program checks for the K nearest neighbours of the test data based on the Eucledian Distance. And out of those K neighbours checks the class with maximum points and assign the test data with the same class.

To do this we first import the essential libraries.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

Since, this project is done on Google Colab, we first have to upload the dataset to the colab server as follows.

In [3]:
from google.colab import files
uploaded = files.upload()

Saving datasets_4247_6570_Iris.csv to datasets_4247_6570_Iris.csv


Then, we make use of the pandas library to read the csv file and set the index to the column 'Id'!

In [5]:
import io
df = pd.read_csv(io.BytesIO(uploaded['datasets_4247_6570_Iris.csv'])).set_index('Id')
df.head()

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
Id                                                                       
1             5.1           3.5            1.4           0.2  Iris-setosa
2             4.9           3.0            1.4           0.2  Iris-setosa
3             4.7           3.2            1.3           0.2  Iris-setosa
4             4.6           3.1            1.5           0.2  Iris-setosa
5             5.0           3.6            1.4           0.2  Iris-setosa

We then perform **Data Wrangling**. We drop any rows with null values if present and then check the shape of the data for future reference.

In [6]:
df.dropna(inplace = True)
df.shape

(150, 5)

Next, to split the data into test and train datasets we make use of the **train_test_split()** function provided by **sklearn.model_selection** package.

We set the test size to be 0.33, i.e., 1/3 rd of the original data.

Then, we also sort the training and testing data on the basis of their 'Id'.

In [7]:
from sklearn.model_selection import train_test_split

dtr, dte = train_test_split(df, test_size = 0.33)

dtr.sort_values('Id', inplace = True)
dte.sort_values('Id', inplace = True)

print(df['Species'].value_counts(),'\n', dtr['Species'].value_counts(),'\n', dte['Species'].value_counts())

Iris-versicolor    50
Iris-setosa        50
Iris-virginica     50
Name: Species, dtype: int64 
 Iris-versicolor    35
Iris-virginica     33
Iris-setosa        32
Name: Species, dtype: int64 
 Iris-setosa        18
Iris-virginica     17
Iris-versicolor    15
Name: Species, dtype: int64


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Now, we define a function **getDistance()**. This function takes a sample from the test data and finds the distance with each of the data in the training column and appends it to a list which is then returned at the end.

In [8]:
def getDistance(a,b):
  d = []
  for i in range(len(a)):
    ss = (((a.iloc[i,0]-b[0])**2)+((a.iloc[i,1]-b[1])**2)+((a.iloc[i,2]-b[2])**2)+((a.iloc[i,3]-b[3])**2))**0.5
    d.append(ss)
  return d

Next, we define a function **getNeighbours()** that calls the **getDistance()** function, and adds the distance array to a new column to the training data. It then siorts the dataframe on the basis of the 'Distance' column and returns the first q rows which are the **'K-Nearest Neighbours'**.

In [9]:
def getNeighbours(a,b,q):
  dist = getDistance(a,b)
  a['Distance'] = dist
  a.sort_values('Distance', inplace = True)
  return a.iloc[:q,:-1]

We also define another function **classification()** to finalize the class of the test data.
We make use of the **idxmax()** function to find the class with maximum occurences.

In [10]:
def classification(a,b,q):
  neigh = getNeighbours(a,b,q)
  return neigh['Species'].value_counts().idxmax()

Finally, we define a function **accuracy()** to calculate the efficiency of our KNN model.

In [11]:
def accuracy(pred, test):
  c = 0
  for i in range(len(pred)):
    if(pred[i]==test.iloc[i]):
      c+=1
  acc = (c/len(pred))*100
  return acc

Now, finally, we take the number of neighbours 'K' as input and then check the classification for all the data in the test data with the train data and print the accuracy of our model in the end.

In [12]:
y_pred = []
k = int(input("Enter the number of nearest neighbors to check with: "))
for i in range(len(dte)):
  y_pred.append(classification(dtr, dte.iloc[i], k))
print("Accuracy Of The KNN model for K = ", k, " is ", accuracy(y_pred, dte.iloc[:,-1]))

Enter the number of nearest neighbors to check with: 8


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


Accuracy Of The KNN model for K =  8  is  96.0


Above, we found out that for the model created, we have a **96% accuracy for K = 8** which is pretty good!

**Thank You!!!**